presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java

```diff
@@ -82,40 +93,68 @@ public static RetryDriver retry()
 
     public final RetryDriver maxAttempts(int maxAttempts)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
     public final RetryDriver exponentialBackoff(Duration minSleepTime, Duration maxSleepTime, Duration maxRetryTime, double scaleFactor)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
     public final RetryDriver onRetry(Runnable retryRunnable)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, Optional.ofNullable(retryRunnable), stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, Optional.ofNullable(retryRunnable), stopRetrying, classifier, retryExecutorService);
     }
 
     public RetryDriver stopRetryingWhen(Predicate<Exception> stopRetrying)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
     public RetryDriver withClassifier(Function<Exception, Exception> classifier)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
+    }
+
+    public RetryDriver withAsyncRetryExecuterService(ListeningScheduledExecutorService executerService)
+    {
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, executerService);
+    }
+
+    public <V> ListenableFuture<V> runAsync(String callableName, Callable<ListenableFuture<V>> callable)
+    {
+        requireNonNull(callableName, "callableName is null");
+        requireNonNull(callable, "callable is null");
+        return runAsyncInternal(callableName, callable, new RetryStatus());
     }
 
     public <V> V run(String callableName, Callable<V> callable)
     {
         requireNonNull(callableName, "callableName is null");
         requireNonNull(callable, "callable is null");
+        return runInternal(callableName, callable, new RetryStatus());
```

Comment:
Sorry for the late reply. 
In order to check the return value type of the `Callable` I ended up implementing the following function:
```java
private boolean isAsyncCallable(Object object)
{
    for (Type genericInterface : object.getClass().getGenericInterfaces()) {
        if (genericInterface instanceof ParameterizedType) {
            ParameterizedType parameterizedType = (ParameterizedType) genericInterface;
            if (parameterizedType.getRawType().equals(Callable.class) && parameterizedType.getActualTypeArguments()[0] instanceof ParameterizedType) {
                ParameterizedType returnValueType = (ParameterizedType) parameterizedType.getActualTypeArguments()[0];
                if (returnValueType.getRawType().equals(ListenableFuture.class)) {
                    return true;
                }
            }
        }
    }
    return false;
}
```
Will it be too complicated to hack into the runtime type here? Shall we just add a javadoc/comment for `run()` and `runAsync()` to explain their usage instead?